### DISCUSSION TITLE
[Feature] JPEG-XL Support

### DISCUSSION THREAD

**Discussion by Adrian-Grimm** [immich-app/immich#12945] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

## Background 

JPEG-XL is ready for getting mainstream - https://jpegxl.info

Just a few reasons:
- Quality is far better than every alternative by a lower file size in every case
- iPhone 16 is able to shoot JPEG-XL by default
- iOS 17+, macOS, and further OS have full support
- Browsers are all getting or have already base support
- ...

## Target

1. Immich should support JPEG-XL photos to store and show such in native JPEG-XL if the platform supports it and photo was saved in that format.
3. If there is no support on an old platform Immich should always convert it seamless to the old JPEG format for compatibility reasons.
4. May this feature should be an opt-out feature which could get deactivated on server side.

## Out-of-Scope

- The creation of previews doesn't need to get modified regarding the target format for easy compatibility as long as JPEG-XL support is not everywhere given by default. Something like 50% of the devices should support it before considering this as an optional feature.

### Platform

- [X] Server
- [X] Web
- [X] Mobile

**Comment by bo0tzz** [comment]
Immich already supports uploading jxl files, what are you asking for exactly?

**Reply by Adrian-Grimm** [reply]
To also show such files in native JPEG-XL if the platform supports it (except of such small previews).

**Reply by mertalev** [reply]
Since that basically means Safari at the moment, it'd be as simple as adding it in [this](https://github.com/immich-app/immich/blob/63ad3c83732eff4fd71be7c8619d3b8ee5e55f2b/web/src/lib/utils/asset-utils.ts#L338) line.

We can present the original JPEG XL image for other browsers when they start supporting it. I've read that Firefox is interested in JPEG XL, but wants a Rust-based decoder for it. I haven't seen an update on the Chromium team's stance following Apple's push for JPEG XL, but maybe it will force their hand at some point.

**Reply by akostadinov** [reply]
btw Chrome 145 already supports it. Maybe deserves a little more love now.

**Comment by alextran1502** [comment]
Not being supported on mainstream clients yet 
<img width="1501" alt="image" src="https://github.com/user-attachments/assets/1404c77c-37fc-4058-b9d9-482b53c20d9e">

**Reply by Adrian-Grimm** [reply]
right for the web browsers  - but still already some important ones support it and it would be also able to be supported in the Apps...

**Reply by Adrian-Grimm** [reply]
would also help to get the adoption to JPEG-XL pushed.

**Reply by AngelaDMerkel** [reply]
As much as I like JXL, Interop rejected it in their last voting round and only Apple supports read and writing the format. Some camera manufacturers outside of Apple including Samsung support it, but in a very limited capacity. When the format is widely support the progressive decode could save a lot of bandwidth and storage in immich by foregoing the need to generate additional thumbnails

**Reply by Adrian-Grimm** [reply]
I think adoption will speed up from now on - was the same game with HEIF - as soon as Apple pushed it ...

**Reply by akostadinov** [reply]
@Adrian-Grimm comment aged well. Chrome 145 supports it already.

**Reply by hofst** [reply]
@alextran1502 Both Firefox and Chrome just formerly announced the intent to ship jpeg xl, current ETA ~EO September
Now is probably a good time to think about Immich support (Personally, I waited for this exact moment to switch my image pipelines from Raw->jpg to Raw->jxl).

**Reply by mibmo** [reply]
> @alextran1502 Both Firefox and Chrome just formerly announced the intent to ship jpeg xl, current ETA ~EO September

Delighted to hear this! The links for anyone else curious:
- Firefox: https://hacks.mozilla.org/2026/08/intent-to-ship-jpeg-xl
- Chrome: https://groups.google.com/a/chromium.org/g/blink-dev/c/-gDojQbDPRI

**Comment by NicholasFlamy** [comment]
I would like to add an almost side-note feature request, Transcoding JPEG to JPEGXL losslessly to save storage space and have an option to revert back to the original JPEGs since JPEGXL claims this as a key feature.
https://jpegxl.info

From FAQ:

Is Lossless JPEG Transcoding really bit-for-bit accurate?
Converting a JPEG to JPEG XL using `cjxl` and then back to JPEG with `djxl` results in a bit-identical copy of the original JPEG image. However, it’s important to note JPEG standard leaves room for some interpretation, leading to potential minor variations between different decoders.

Do other JPEG XL tools support Lossless JPEG Transcoding?
Other encoders may not support lossless transcoding, so it’s best to check any support documention to verify. Note that lossless encoding of a JPEG file to JPEG XL is not the same as lossless transcoding. Transcoding uses a specific mode to achieve its 20% size gain while maintaining pixel-level accuracy.

**Reply by AngelaDMerkel** [reply]
I suggested this in another feature request but it was closed as not being within the project scope

**Reply by NicholasFlamy** [reply]
Yep, I see #8907.

You were talking about compression in general too.

One thing I will add to my comment is, having this as an immich feature could make it possible to store the images as JPEGXL and on the fly convert to the original JPEG losslessly for clients not supporting JPEGXL. (And what I mentioned about being able to download the original JPEG as though it was never converted.)

**Reply by NicholasFlamy** [reply]
What I'm suggesting might still be out of the scope of the project, but maybe they would be okay with someone like me implementing it and PRing.

That or they give more addon options such that these features could be an addon, because the Lossless Transcode to JPEGXL could already be done by a separate tool, and clients can use the thumbnail and preview images, but if you hit load original image or you zoom in, but your client doesn't support JPEGXL, then on the fly transcoding to JPEG losslessly would be beneficial. So some integration for that and a way to download the losslessly converted back to JPEG image is needed.

**Reply by bo0tzz** [reply]
We have no intent to add any changing of original files to Immich, but the idea is to make it easy to do this via a community addon at some point yes.

**Comment by Qhilm** [comment]
>iPhone 16 is able to shoot JPEG-XL by default

iPhone 16 shoots DNG with a JPEG XL _inside_ the DNG container, that is _not_ a JPEG XL file, just to be clear, it's a DNG file.

Secondly, it is for ProRAW pictures _only_ and that is _not_ the default.

I am big fan of JPEG XL, but pushing it based on inaccurate facts is counterproductive.

We need Chrome to support it, that would move the tides.

**Reply by Adrian-Grimm** [reply]
You are right, but DNG is just a container and iPhone also support it without the container. But it would be best to support both cases - with and without the DNG container.

**Reply by jarrodmoldrich** [reply]
> We need Chrome to support it, that would move the tides.

For Immich—personally—I'd only need Firefox to support it.  That way all platforms _can_ view  natively.  At this point jxl could be the only image format I'd need, and Chrome browsers could be redirected at the reverse proxy level to a helpful message to use Firefox instead.

**Reply by NicholasFlamy** [reply]
> > We need Chrome to support it, that would move the tides.
> 
> For Immich personally I'd only need Firefox to support it. That way all platforms _can_ view natively. At this point jxl could be the only image format I'd need, and Chrome browsers could be redirected at the reverse proxy level to a helpful message to use Firefox instead.

Haha, while that's funny, I personally use Brave because it's open source and you can self-host a sync server while also using Chromium. I need a Chromium-based browser for HEVC (AV1 still isn't old enough and many things record and support HEVC, my Galaxy S22 has a HW HEVC decoder but no HW AV1 decoder) so for many, that's not an option.

**Reply by NicholasFlamy** [reply]
> We need Chrome to support it, that would move the tides.

I'm pretty sure Chrome actually REMOVED support a year or two ago. I think it was political as they then pushed WebP and AVIF (both of which are worse than JXL in just about every way).

**Reply by jarrodmoldrich** [reply]
Yeah, I get that other peoples' needs don't match mine 🙂  But man, I'd just love to economise my server storage without reducing image quality.  I really hate Google for this.

> my Galaxy S22 has a HW HEVC decoder but no HW AV1 decoder

How bad is the CPU usage of the software decoder?

I'm thinking I might just convert all my images to AVIF.  But I don't know if under the hood Immich is just converting them to jpg anyway.

**Reply by NicholasFlamy** [reply]
> But I don't know if under the hood Immich is just converting them to jpg anyway.

There is a thumbnail and preview option that stores a converted COPY. Immich DOES NOT alter your original photos.

**Reply by NicholasFlamy** [reply]
> I really hate Google for this.

Me too. Join the club.

> How bad is the CPU usage of the software decoder?

I haven't tested this, but I know it's more than HW decoding for sure. Based on my Ryzen 5 5600X, it's probably pretty bad, luckily my Radeon RX 6700XT has HW AV1 decode.

**Reply by jarrodmoldrich** [reply]
> I haven't tested this, but I know it's more than HW decoding for sure. Based on my Ryzen 5 5600X, it's probably pretty bad, luckily my Radeon RX 6700XT has HW AV1 decode.

Went into a rabbit hole on this.  Seems that at 32 QP @ 1080p 30/60fps, the March 2022 version of dav1d uses close to 3 cores on a selection of Android devices (from 2018-2020).  Seems like most recent phones can handle it, but they might get a bit warm.

https://www.linkedin.com/feed/update/urn:li:activity:6968047760256241664/
![image](https://github.com/user-attachments/assets/ecd46aaa-de3b-43a2-80cc-69e0b516a654)

> There is a thumbnail and preview option that stores a converted COPY. Immich DOES NOT alter your original photos.

So previews are what are seen on the app/web when you click on the picture?  Does that mean that all images, even JPGs, are copied?  If so, then I'll chose `jxl` just to reduce on server space, though it would be nice if it used it for display as well.

**Reply by NicholasFlamy** [reply]
> Seems like most recent phones can handle it, but they might get a bit warm.

And that heat is from electricity and that electricity is from the 3700 mAH battery in my phone. My phone doesn't last long enough as it is. I've considered the amazing Asus Zenphone 10 but they gave it only two years of software update and are such sh*tholes that they got rid of bootloader unlocking while the Zenphone 9 was still for sale.

**Reply by jarrodmoldrich** [reply]
@NicholasFlamy Just reposting this question to you because I buried it a bit in the last one

> There is a thumbnail and preview option that stores a converted COPY. Immich DOES NOT alter your original photos.

So previews are what are seen on the app/web when you click on the picture? Does that mean that all images, even JPGs, are copied? If so, then I'll chose jxl just to reduce on server space, though it would be nice if it used it for display as well.

**Reply by NicholasFlamy** [reply]
> So previews are what are seen on the app/web when you click on the picture? 

Yes with the default setting of show originals turned off. (When you zoom in it'll load the original.)

> Does that mean that all images, even JPGs, are copied?

Yes, you can select JPEG or WebP.

**Reply by jarrodmoldrich** [reply]
> (When you zoom in it'll load the original.)

Right!  So the timeline/album view are the thumbnail copies.  When you click on the image it shows the preview copy, and when zooming it shows the original.  That makes sense.  Wish that was more obvious or documented somewhere.  So I guess if you zoom into a jxl on an unsupported browser it just shows the preview.  Which I'm okay with.

**Comment by aktau** [comment]
Chrome has merged Jpeg XL (16 Jan 2026): https://news.ycombinator.com/item?id=46597927. AFAICT, it's currently available in Chrome Beta.

**Reply by yodatak** [reply]
Some more benchmark of why jpgxl is awsome for photography https://cloudinary.com/blog/jpeg-xl-and-the-pareto-front

**Reply by Qhilm** [reply]
I really hope Chrome will support it with HDR. Safari already does.

HDR is THE big upside of JPEG XL here. If we can have HDR previews and thumbnails it the browser and in the app, that would be _so_ awesome.

**Reply by Qhilm** [reply]
@NicholasFlamy: you don't like HDR in general?

**Reply by NicholasFlamy** [reply]
> @NicholasFlamy: you don't like HDR in general?

HDR is not THE big upside. JPEG-XL is just plain better. HDR is just one of many improvements. The majority of people don't even have HDR monitors.

**Reply by Qhilm** [reply]
well yeah, every smartphone is an HDR monitor, I would argue that's the majority of screens out there.

JPEG XL is indeed better in a lot of respects, but from a picture quality point of view, the improvement would be HDR. HDR makes the picture look more realistic.

**Comment by ranidspace** [comment]
JXL is now supported in a chrome stable release, though under a flag which needs to be manually enabled.

[Firefox has merged a replacement for their old JXL decoder](https://github.com/mozilla-firefox/firefox/commits/main/image/rust/jxl), and may be included in firefox 150, based on when the next nightly builds start.

**Comment by nicosemp** [comment]
Immich support for JPEG XL is live since last year for Safari only. I recently added a more generic support detection, released in [2.5.6](https://github.com/immich-app/immich/releases/tag/v2.5.6). There's no description but it links to PR #25599 in the New Contributors section.

I'm working on a more fine-grained detection in #26122:
- All new browsers that support JPEG XL images will be automatically detected, without further code changes.
  - Both stable Chrome and Firefox (EDIT: only Firefox nightly for now) have flags to enable JPEG XL support.
  - Zen Browser (based on Firefox) has the flag enabled by default.
- Only Safari `>=17` should return true for JPEG XL support (Safari `<17` also returns true now).

**Reply by ranidspace** [reply]
The firefox flag only exists on nightly builds, however the firefox JXL decoder is being replaced so that may change.

Is there any support possible for mobile?

**Reply by nicosemp** [reply]
> The firefox flag only exists on nightly builds

True, hopefully Firefox adds the flag to stable soon.

Mobile support should already be live for iOS since version 17, together with Safari. For Android I guess `libjxl-tools` could be used to manually add support. However I guess maintainers aren't going to spend time on something that will eventually be implemented at the OS level.

**Reply by ranidspace** [reply]
I really hope chrome/firefox also enables the flag by default. I've checked out the chromium implementation and it works perfectly for the few things I've seen. Most people won't ever go and enable any flags at all. I could complain more but there's more productive places to do it

**Reply by NicholasFlamy** [reply]
> I really hope chrome/firefox also enables the flag by default. I've checked out the chromium implementation and it works perfectly for the few things I've seen. Most people won't ever go and enable any flags at all. I could complain more but there's more productive places to do it

I hope you know that things like new decoders and stuff usually start out as a flag default off, then default on once they know it works great and then in the future when there's no issues they remove the flag.

Edit: TLDR have patience

**Reply by Qhilm** [reply]
To be fair, there used to be a flag 4 years ago on Chrome and it got removed. Cautiously optimistic that this time it will become the default, but it will take probably another year.

Once it is default, clients will have to update but down the line (another couple of years probably), at some point, a majority of android devices will be able to render jpeg xl. Hopefully including HDR.

I would find it really great if JPEG XL support (including HDR, including thumbnails and previews) could be added before then. I have to admit, I don't know how much effort it is, it could be high if Immich has to introduce a new library (libvips?) and consolidate existing image processing/conversion to libvips.

**Reply by mertalev** [reply]
Adding it for images is straightforward; it's just a matter of setting the target format to 'jxl' for sharp since it's already built with libjxl. Our ffmpeg build probably doesn't have libjxl though.

I should also mention that I don't think the JXL output will be HDR this way unless libvips handles this.

**Reply by nicosemp** [reply]
So you guys would consider adding JPEG XL support to the Android app too, without waiting for OS support?

That would be great, as it would take way more than 2 years to have a meaningful percentage of Android devices receive support, due to version fragmentation.

**Reply by mertalev** [reply]
We would have to look into what Android support without relying on the OS would entail. It's probably feasible, but I can't make promises. In any case, I'm fine with adding JXL as a target format before it becomes completely baseline.

**Reply by AngelaDMerkel** [reply]
@mertalev [libvips 1.18 has support for gainsmaps](https://github.com/libvips/libvips/releases/tag/v8.18.0) and these should be preserved when converting from an ultrahdr JPEG to JXL or vice versa.

**Reply by mertalev** [reply]
It does and I could be wrong, but I was under the impression it can only target Ultra HDR JPEG and other formats are still SDR. It might need [this](https://github.com/libvips/libvips/pull/4818) PR to handle HDR for other formats (though this seems to only tackle the input file rather than output).

**Reply by Zarbizaure** [reply]
Waterfox (Firefox-based) also now supports it by default, both on its desktop and mobile versions.
I 100% support an early support of this format in immich; just waiting it to shift my RAW processing from JPEG to JXL output !

**Reply by Qhilm** [reply]
@Zarbizaure what is stopping you from shifting to JPEG XL? I already export to jxl today, it works fine in Immich.

Thumbnails and previews are not jpeg xl but they are correctly generated.

HDR rendering is a different story though.

**Reply by Starbix** [reply]
The PR will also handle output: both an output with CICP signals or a generated gainmap is supported (libvips can generate gainmaps from images in scRGB colour space). The JXL output isn't fully hooked up yet in the PR as JXL doesn't implement a 1:1 match of CICP values. That's not a huge problem, we only need to decide what happens in case of a mismatch (most likely fallback to sRGB).

**Comment by ppskj178** [comment]
I’ve initiated a separate discussion ([#28423](https://github.com/immich-app/immich/discussions/28423)) regarding the aggressive use of JPEG XL's progressive and ROI decoding to serve all thumbnails and previews from a single master image. This architectural shift could significantly optimize storage and bandwidth by eliminating the need for multiple generated preview files and face crops. While currently premature due to limited browser support, I hope this vision can be considered as part of Immich's long-term JPEG XL roadmap.

**Reply by nicosemp** [reply]
It will [probably](https://github.com/immich-app/immich/discussions/26399) be closed as a duplicate of this. Doesn't matter if you have a different scope in mind, apparently anything pertaining to JPEG XL belongs to this one discussion only.

**Reply by ppskj178** [reply]
@nicosemp 

My proposal goes beyond simply adding support for JPEG XL; it also ties directly into Immich's architecture. If my post is merged into this single thread, I cannot even begin to imagine how long it would take to achieve full JPEG XL support.

Fortunately, contrary to your concern, a maintainer recently left a comment on my post, and they did not mention anything about it being a duplicate.

**Reply by nicosemp** [reply]
I'm 100% with you on the proposal. JPEG XL has many advantages, and Immich could make good use of it beyond just supporting it as an additional mime type. Sorry to see that the idea is being deferred just because JPEG XL is not on Chrome/Firefox stable yet.

**Reply by ppskj178** [reply]
@nicosemp

Indeed, many recent image formats have clear limitations as they are derived from video codecs. JPEG XL is free from these constraints and is definitely worth researching further.

It’s quite disappointing to see my request closed. As the maintainer suggested, I truly hope this proposal will be revisited in the future.

**Reply by Qhilm** [reply]
We should list the different sub-elements:

1. ~~support importing JPEG XL files~~ ✅ done.
2. ~~support importing JPEG XL HDR files~~  ✅ done, Previews and Thumbnails of JPEG XL HDR masters are correctly tone mapped.
3. Decide if tone mapping should be left to the client or if Immich should detect if HDR is supported (I favour the latter, tone mapping results of clients are fairly unpredictable)
4. [Maybe] Implement detectionf or HDR support
5. Implement JPEG XL support for thumbnails – with HDR support, depending on 3/
6. Implement JPEG XL support for previews – with HDR support, depending on 3/
7. Implement consolidated preview+thumbnail in one file (this is elegant but really a nice to have compared to the effort involved in doing this, in my view)

Did I forget something? There's a lot...

**Reply by mertalev** [reply]
The approach for JPEG-XL is generally native HDR with client tone-mapping. It does support gain maps as an alternative, but this is less common and I don't think the image libraries support encoding it.

I looked into how viable the progressive decoding approach would be some time ago. One thing is that decoding a partial progressive image required decoding at least twice the size of the WebP thumbnails before it became comparable in quality. Another is that browsers don't necessarily even support progressive decoding for JPEG-XL yet; the mobile app definitely doesn't. I think JPEG-XL is a great format and adding support for it would be good, but it's premature at best to rely on it beyond that.

**Reply by ppskj178** [reply]
@mertalev

I'm impressed that you've already looked into this! I was aware that progressive decoding has high memory and CPU usage, but I didn't realize there were also issues with quality relative to file size. It definitely seems like there are quite a few challenges that need to be verified and resolved.

While it might be premature right now, I'll be looking forward to the day when it becomes feasible and can be reconsidered. Thank you as always for your hard work!

**Reply by AngelaDMerkel** [reply]
@mertalev While progressive decode is likely not an ideal solution for thumbnails, I believe that it should be supported for viewing the full image. 100MP images can take some time to load, and it would be a good compromise between generating previews vs originals in the absence of some tiling solution.

**Reply by mertalev** [reply]
@AngelaDMerkel Yes, the intent is for it to respect the progressive option in the admin settings.

**Comment by frenzybiscuit** [comment]
What's the status on this (Jpeg-xl preview images) ?

I use safari, but I don't see a JXL preview option for the extension.

**Reply by nicosemp** [reply]
For now only native JPEG XL images will be shown on Safari, and only if you are looking at the original image.

There is no thumbnail/preview/full-size JPEG XL generation yet. They'll probably wait until it gets released on stable Chrome/Firefox without flags.

**Reply by dyptan-io** [reply]
Thumbnail/preview JPEG XL generation will help us displaying HDR in pictures I think, since neither JPEG nor WebP support it.

**Reply by mertalev** [reply]
Kind of. Ultra HDR JPEGs exist, which start with an SDR base layer for compatibility and a gain map to boost the SDR base when the client supports it.

The same technique exists for JPEG XL as well, but JPEG XL can also be native HDR without an SDR base. Compared to gain maps, this way produces better colors for HDR clients, but worse results for SDR.

**Reply by dyptan-io** [reply]
Thanks for explaining it, @mertalev! HDR is kind of standard today at mobile photos, and I refuse to see my photos in SDR. Are there any obstacles left in implementing it full stack - mobile, web, etc.?

**Reply by Starbix** [reply]
Large parts of it depend on the next libvips version (v8.19) after which we'll need to wait for sharp v0.36. Then immich can start with a better HDR pipeline. There's no roadmap for libvips, but maybe it'll happen until autumn this year.

**Reply by mertalev** [reply]
The libvips part of it is for previews, though. Rendering the originals in HDR isn't blocked on libvips. A lot of the work is in being able to display HDR in the app at all due to Flutter limitations. If we can make it work well for originals, then HDR previews would work as well after libvips 8.19 releases.

**Comment by or-dodson** [comment]
Both [Firefox](https://groups.google.com/a/mozilla.org/g/dev-platform/c/3YMV4MS34KA/m/iqfJV5cXEQAJ) and [Chrome](https://groups.google.com/a/chromium.org/g/blink-dev/c/-gDojQbDPRI) have issued an intent to ship JPEG-XL decoding support by default on all platforms using jxl-rs. This library supports progressive decoding, wide color gamut, HDR, and animation. The current timeline to add support in each browser is sometime in November with Firefox 157 and [Chrome 155](https://chromestatus.com/feature/5114042131808256?gate=6594362739916800).

With this, almost all clients will support jpeg xl viewing by default. When that day comes, setting sensible default settings, waiting for library support, and implementing details will be the last steps for full support of jpeg xl in Immich.
